src/peergos/shared/crypto/FragmentedPaddedCipherText.java

```diff
@@ -0,0 +1,136 @@
+    private final List<Multihash> cipherTextFragments;
+
+    public FragmentedPaddedCipherText(byte[] nonce, List<Multihash> cipherTextFragments) {
+        this.nonce = nonce;
+        this.cipherTextFragments = cipherTextFragments;
+    }
+
+    @Override
+    public CborObject toCbor() {
+        SortedMap<String, Cborable> state = new TreeMap<>();
+        state.put("n", new CborObject.CborByteArray(nonce));
+        state.put("f", new CborObject.CborList(cipherTextFragments
+                        .stream()
+                        .map(CborObject.CborMerkleLink::new)
+                        .collect(Collectors.toList())));
+        return CborObject.CborMap.build(state);
+    }
+
+    public static FragmentedPaddedCipherText fromCbor(Cborable cbor) {
+        if (! (cbor instanceof CborObject.CborMap))
+            throw new IllegalStateException("Incorrect cbor for FragmentedPaddedCipherText: " + cbor);
+
+        CborObject.CborMap m = (CborObject.CborMap) cbor;
+
+        byte[] nonce =  m.getByteArray("n");
+        List<Multihash> fragmentHashes = m.getList("f").value
+                .stream()
+                .map(c -> ((CborObject.CborMerkleLink)c).target)
+                .collect(Collectors.toList());
+        return new FragmentedPaddedCipherText(nonce, fragmentHashes);
+    }
+
+    protected static byte[] pad(byte[] input, int blockSize) {
+        int nBlocks = (input.length + blockSize - 1) / blockSize;
+        return Arrays.copyOfRange(input, 0, nBlocks * blockSize);
+    }
+
+    public static <T extends Cborable> Pair<FragmentedPaddedCipherText, List<FragmentWithHash>> build(SymmetricKey from,
+                                                                                                      T secret,
+                                                                                                      int paddingBlockSize,
+                                                                                                      int maxFragmentSize,
+                                                                                                      Hasher hasher) {
+        if (paddingBlockSize < 1)
+            throw new IllegalStateException("Invalid padding block size: " + paddingBlockSize);
+        byte[] nonce = from.createNonce();
+        byte[] cipherText = from.encrypt(pad(secret.serialize(), paddingBlockSize), nonce);
+
+        if (cipherText.length <= 4096 + TweetNaCl.SECRETBOX_OVERHEAD_BYTES) {
+            // use inline identity hash for small amount of data (small files or directories)
+            FragmentWithHash frag = new FragmentWithHash(new Fragment(cipherText), hasher.identityHash(cipherText, true));
+            return new Pair<>(new FragmentedPaddedCipherText(nonce, Collections.singletonList(frag.hash)), Collections.singletonList(frag));
+        }
+
+        byte[][] split = split(cipherText, maxFragmentSize);
+
+        List<FragmentWithHash> frags = Arrays.asList(split).stream()
```

Comment:
nit: ``Stream.of(spilt)``

Reply:
Cheers. I've changed it to Arrays.stream() (in the second PR)